src/lib/statistics/statistics_objects/range_filter.cpp

```diff
@@ -0,0 +1,288 @@
+                                                             uint32_t max_ranges_count) {
+  static_assert(std::is_arithmetic_v<T>, "Range filters are only allowed on arithmetic types.");
+  DebugAssert(max_ranges_count > 0, "Number of ranges to create needs to be larger zero.");
+  DebugAssert(std::is_sorted(dictionary.begin(), dictionary.cend()), "Dictionary must be sorted in ascending order.");
+
+  if (dictionary.empty()) {
+    // Empty dictionaries will, e.g., occur in segments with only NULLs - or empty segments.
+    return std::make_unique<RangeFilter<T>>(std::vector<std::pair<T, T>>{});
+  }
+
+  if (dictionary.size() == 1) {
+    std::vector<std::pair<T, T>> ranges;
+    ranges.emplace_back(dictionary.front(), dictionary.front());
+    return std::make_unique<RangeFilter<T>>(std::move(ranges));
+  }
+
+  /*
+  * In case more than one value is present, first the elements are checked for potential overflows (e.g., when calculating
+  * the distince between INT::MIN() and INT::MAX(), the resulting distance might be to large for signed types).
+  * While being rather unlikely for doubles, it's more likely to happen when Opossum includes tinyint etc.
+  * std::make_unsigned<T>::type would be possible to use for signed int types, but not for floating types.
+  * Approach: take the min and max values and simply check if the distance between both might overflow. In this case,
+  * fall back to a single range filter.
+  */
+  const auto min_max = std::minmax_element(dictionary.cbegin(), dictionary.cend());
+  if ((*min_max.first < 0) &&
+      (*min_max.second > std::numeric_limits<T>::max() + *min_max.first)) {  // min_value is negative
+    return std::make_unique<RangeFilter<T>>(std::vector<std::pair<T, T>>{{*min_max.first, *min_max.second}});
+  }
+
+  // calculate distances by taking the difference between two neighbouring elements
+  // vector stores <distance to next element, dictionary index>
+  std::vector<std::pair<T, size_t>> distances;
+  distances.reserve(dictionary.size() - 1);
+  for (auto dict_it = dictionary.cbegin(); dict_it + 1 != dictionary.cend(); ++dict_it) {
+    auto dict_it_next = dict_it + 1;
+    distances.emplace_back(*dict_it_next - *dict_it, std::distance(dictionary.cbegin(), dict_it));
+  }
+
+  std::sort(distances.begin(), distances.end(),
+            [](const auto& pair1, const auto& pair2) { return pair1.first > pair2.first; });
+
+  if ((max_ranges_count - 1) < distances.size()) {
+    distances.erase(distances.cbegin() + (max_ranges_count - 1), distances.cend());
+  }
+
+  std::sort(distances.begin(), distances.end(),
+            [](const auto& pair1, const auto& pair2) { return pair1.second < pair2.second; });
+  // we want a range until the last element in the dictionary
+  distances.emplace_back(T{}, dictionary.size() - 1);
+
+  // derive intervals from distances where items exist
+  //
+  // start   end  next_startpoint
+  // v       v    v
+  // 1 2 3 4 5    10 11     15 16
```

Comment:
I think this would be clearer if the example would run like a golden thread through this. Reading this from the start, I had to think for a while why you sort the distances etc.